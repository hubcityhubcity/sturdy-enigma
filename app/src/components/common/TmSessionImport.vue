<template>
  <SessionFrame :icon="`settings_backup_restore`">
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title">
        Recover with backup code
      </h2>
      <div class="session-main bottom-indent">
        <Steps :steps="[`Recover`, `Name`, `Password`]" active-step="Recover" />
        <TmFormGroup
          :error="$v.$error && $v.seed.$invalid"
          field-id="import-seed"
          field-label="Backup code"
        >
          <FieldSeed
            id="import-seed"
            :value="seed"
            :placeholder="
              isPolkadot
                ? 'Must be your seed phrase or private key hash'
                : 'Must be exactly 12 or 24 words'
            "
            @input="val => (seed = val)"
          />
          <div class="polkadot-algorithm-div">
            <span>{{ polkadotAlgo }}</span>
            <label v-if="isPolkadot" class="polkadot-algorithm-switch">
              <input
                v-model="polkadotAlgoCheckbox"
                type="checkbox"
                @change="switchPolkadotAlgorithm()"
              />
              <span class="slider round"></span>
            </label>
          </div>
          <TmFormMsg
            v-if="$v.seed.$error && !$v.seed.required"
            name="Seed"
            type="required"
          />
          <TmFormMsg
            v-else-if="$v.seed.$error && !$v.seed.seedHasCorrectLength"
            name="Seed"
            :type="isPolkadot ? 'incorrectPolkadotSeed' : 'words12or24'"
          />
          <TmFormMsg
            v-else-if="$v.seed.$error && !$v.seed.seedIsLowerCaseAndSpaces"
            name="Seed"
            :type="isPolkadot ? 'incorrectPolkadotSeed' : 'lowercaseAndSpaces'"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Next" type="submit" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmFormMsg from "common/TmFormMsg"
import FieldSeed from "common/TmFieldSeed"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
import Steps from "../../ActionModal/components/Steps"
import { mnemonicValidate } from "@polkadot/util-crypto"

const has12or24words = (param) => {
  return (
    param && (param.split(` `).length === 12 || param.split(` `).length === 24)
  )
}
const isHex = (value) => {
  return /0x[0-9a-f]+/i.test(value)
}

const lowerCaseAndSpaces = (param) => {
  const seedWordsAreLowerCaseAndSpaces = /^([a-z]+\s)*[a-z]+$/g
  if (param.match(seedWordsAreLowerCaseAndSpaces)) {
    return param === param.match(seedWordsAreLowerCaseAndSpaces)[0]
  }
  return false
}
// exporting these for testing
export const isPolkadotHexSeed = (seed) => {
  return isHex(seed) && seed.length === 66
}

export const polkadotRawSeedValidate = (seed) => {
  return (seed.length > 0 && seed.length <= 32) || isPolkadotHexSeed(seed)
}

export const polkadotValidation = (seed) => {
  return mnemonicValidate(seed) || polkadotRawSeedValidate(seed)
}

export default {
  name: `session-import-backupcode`,
  components: {
    TmBtn,
    SessionFrame,
    FieldSeed,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    Steps,
  },
  data: () => {
    return {
      polkadotAlgo: "Schnorrkel",
      polkadotAlgoCheckbox: false,
      polkadotAlgoDictionary: {
        Schnorrkel: "sr25519",
        Edwards: "ed25519"
      }
    }
  },
  computed: {
    ...mapGetters([`recover`, `currentNetwork`]),
    seed: {
      get() {
        return this.$store.state.recover.seed
      },
      set(value) {
        this.$store.commit(`updateField`, {
          field: `seed`,
          value: value.trim(), // remove spaces from beginning and end of string
        })
      },
    },
    isPolkadot() {
      return this.currentNetwork.network_type === "polkadot"
    },
  },
  methods: {
    onSubmit() {
      this.$v.$touch()
      if (this.$v.seed.$invalid || this.$v.seed.$invalid) return
      this.$router.push("/recover/name")
    },
    switchPolkadotAlgorithm() {
      this.polkadotAlgo = this.polkadotAlgoCheckbox ? "Edwards" : "Schnorrkel"
      this.$store.dispatch(
        "setPolkadotAlgo",
        this.polkadotAlgoDictionary[this.polkadotAlgo]
      )
    }
  },
  validations() {
    return {
      seed: {
        required,
        seedIsLowerCaseAndSpaces: (param) =>
          this.isPolkadot
            ? polkadotValidation(param)
            : lowerCaseAndSpaces(param),
        seedHasCorrectLength: (param) =>
          this.isPolkadot ? polkadotValidation(param) : has12or24words(param),
      },
    }
  },
}
</script>
<style scoped>
.polkadot-algorithm-div {
  margin-top: 2vh;
}

.polkadot-algorithm-div span {
  display: block;
  font-style: italic;
  font-size: 1.7vh;
  color: var(--primary);
  font-weight: 500;
  margin-bottom: 1vh;
}

.polkadot-algorithm-switch {
  position: relative;
  display: block;
  width: 11vw;
  height: 6vh;
}

.polkadot-algorithm-switch input {
  opacity: 0;
  width: 0;
  height: 0;
}

.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: var(--primary);
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.slider:before {
  position: absolute;
  content: "";
  height: 26px;
  width: 26px;
  left: 4px;
  bottom: 4px;
  background-color: white;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

input:checked + .slider {
  background-color: var(--primary-hover);
}

input:focus + .slider {
  box-shadow: 0 0 1px var(--primary-hover);
}

input:checked + .slider:before {
  -webkit-transform: translateX(26px);
  -ms-transform: translateX(26px);
  transform: translateX(26px);
}

.slider.round {
  border-radius: 34px;
}

.slider.round:before {
  border-radius: 50%;
}
</style>
