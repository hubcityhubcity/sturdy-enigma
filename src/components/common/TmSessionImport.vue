<template>
  <SessionFrame>
    <TmFormStruct :submit="onSubmit.bind(this)">
      <h2 class="session-title">
        Recover with backup code
      </h2>
      <div class="session-main">
        <TmFormGroup
          :error="$v.$error && $v.fields.importName.$invalid"
          field-id="import-name"
          field-label="Account Name"
        >
          <TmField
            id="import-name"
            v-model.trim="fields.importName"
            type="text"
            placeholder="Must have at least 5 characters"
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.fields.importName.$error && !$v.fields.importName.required"
            name="Name"
            type="required"
          />
          <TmFormMsg
            v-if="
              $v.fields.importName.$error && !$v.fields.importName.minLength
            "
            name="Name"
            type="minLength"
            min="5"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.fields.importPassword.$invalid"
          field-id="import-password"
          field-label="Password"
        >
          <TmField
            id="import-password"
            v-model="fields.importPassword"
            type="password"
            placeholder="Must be at least 10 characters"
          />
          <TmFormMsg
            v-if="
              $v.fields.importPassword.$error &&
                !$v.fields.importPassword.required
            "
            name="Password"
            type="required"
          />
          <TmFormMsg
            v-if="
              $v.fields.importPassword.$error &&
                !$v.fields.importPassword.minLength
            "
            name="Password"
            type="minLength"
            min="10"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.fields.importPasswordConfirm.$invalid"
          field-id="import-password-confirmation"
          field-label="Confirm Password"
        >
          <TmField
            id="import-password-confirmation"
            v-model="fields.importPasswordConfirm"
            type="password"
            placeholder="Enter password again"
          />
          <TmFormMsg
            v-if="
              $v.fields.importPasswordConfirm.$error &&
                !$v.fields.importPasswordConfirm.sameAsPassword
            "
            name="Password confirmation"
            type="match"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.$error && $v.fields.importSeed.$invalid"
          field-id="import-seed"
          field-label="Backup code"
        >
          <FieldSeed
            id="import-seed"
            :value="fields.importSeed"
            placeholder="Must be exactly 24 words"
            @input="val => (fields.importSeed = val)"
          />
          <TmFormMsg
            v-if="$v.fields.importSeed.$error && !$v.fields.importSeed.required"
            name="Seed"
            type="required"
          />
          <TmFormMsg
            v-else-if="
              $v.fields.importSeed.$error && !$v.fields.importSeed.words24
            "
            name="Seed"
            type="words24"
          />
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Recover" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { required, minLength, sameAs } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import FieldSeed from "common/TmFieldSeed"
import SessionFrame from "common/SessionFrame"
import { mapGetters } from "vuex"
export default {
  name: `session-import`,
  components: {
    TmBtn,
    TmField,
    SessionFrame,
    FieldSeed,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct
  },
  data: () => ({
    fields: {
      importName: ``,
      importPassword: ``,
      importPasswordConfirm: ``,
      importSeed: ``
    }
  }),
  computed: {
    ...mapGetters([`connected`])
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        await this.$store.dispatch(`createKey`, {
          seedPhrase: this.fields.importSeed,
          password: this.fields.importPassword,
          name: this.fields.importName
        })
        this.$router.push(`/`)
      } catch (error) {
        this.$store.commit(`notifyError`, {
          title: `Couldn't create account`,
          body: error.message
        })
      }
    }
  },
  validations: () => ({
    fields: {
      importName: { required, minLength: minLength(5) },
      importPassword: { required, minLength: minLength(10) },
      importPasswordConfirm: { sameAsPassword: sameAs(`importPassword`) },
      importSeed: { required, words24 },
      errorCollection: false
    }
  })
}
const words24 = param => {
  return param && param.split(` `).length === 24
}
</script>
