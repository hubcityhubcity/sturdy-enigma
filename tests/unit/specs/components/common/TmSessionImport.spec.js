import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import TmSessionImport from "common/TmSessionImport"
jest.mock(`scripts/google-analytics.js`, () => () => {})
const seed = `goose toward escape engine wheel board help torch avocado educate rose rebel rigid side aspect abandon grace admit inherit female grant pledge shine inquiry`
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})

describe(`TmSessionImport`, () => {
  let wrapper, $store

  beforeEach(() => {
    const getters = {
      connected: () => true
    }
    $store = {
      getters,
      commit: jest.fn(),
      dispatch: jest.fn(async () => true)
    }
    wrapper = shallowMount(TmSessionImport, {
      localVue,
      mocks: {
        $store,
        $router: {
          push: jest.fn()
        }
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show error if seed is not filled in`, async () => {
    wrapper.setData({ fields: { importSeed: `` } })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show error if seed is 16 words long`, async () => {
    wrapper.setData({
      fields: {
        importSeed: `asdf asdf asdf asdf`
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show error if password is not confirmed`, async () => {
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `notthesame`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should not continue if creation failed`, async () => {
    $store.dispatch = jest.fn(() => Promise.reject(new Error(`Wrong password`)))
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit).toHaveBeenCalledWith(`notifyError`, {
      title: `Couldn't create account`,
      body: expect.stringContaining(`Wrong password`)
    })
  })

  it(`should show a notification if creation failed`, async () => {
    $store.dispatch = jest.fn(() => Promise.reject({ message: `test` }))
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    await wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[0][0]).toEqual(`notifyError`)
    expect($store.commit.mock.calls[0][1].body).toEqual(`test`)
  })

  it(`should go to the home page if recovering is successful`, async () => {
    wrapper.setData({
      fields: {
        importName: `foo123`,
        importPassword: `1234567890`,
        importPasswordConfirm: `1234567890`,
        importSeed: seed
      }
    })
    $store.dispatch = jest.fn(() => Promise.resolve())
    await wrapper.vm.onSubmit()
    expect($store.dispatch.mock.calls[0][0]).toEqual(`createKey`)
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith(`/`)
  })
})
