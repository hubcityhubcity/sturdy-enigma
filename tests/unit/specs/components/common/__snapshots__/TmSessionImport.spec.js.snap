// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionImport has the expected html structure 1`] = `
<sessionframe-stub>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      
      Recover with backup code
    
    </h2>
     
    <div
      class="session-main"
    >
      <tmformgroup-stub
        fieldid="import-name"
        fieldlabel="Account Name"
      >
        <tmfield-stub
          id="import-name"
          placeholder="Must have at least 5 characters"
          type="text"
          value=""
          vue-focus="vue-focus"
        />
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-password"
        fieldlabel="Password"
      >
        <tmfield-stub
          id="import-password"
          placeholder="Must be at least 10 characters"
          type="password"
          value=""
        />
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-password-confirmation"
        fieldlabel="Confirm Password"
      >
        <tmfield-stub
          id="import-password-confirmation"
          placeholder="Enter password again"
          type="password"
          value=""
        />
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="import-seed"
        fieldlabel="Backup code"
      >
        <fieldseed-stub
          id="import-seed"
          placeholder="Must be exactly 24 words"
          value=""
        />
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        value="Recover"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
